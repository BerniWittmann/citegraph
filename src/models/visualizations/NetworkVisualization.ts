import Visualization, { VisualizationBaseFields } from './Visualization'

export interface NetworkVisualizationParameters {
  pruneLeastCitedPercentage: number
  bibliographicCouplingWeight: number
}

export interface NetworkVisualizationDataNode {
  id: string
  weight: number
  color: string
  x: number
  y: number
}

export interface NetworkVisualizationDataEdge {
  source: string
  target: string
}

export interface NetworkVisualizationData {
  data: {
    nodes: Array<NetworkVisualizationDataNode>
    edges: Array<NetworkVisualizationDataEdge>
  }
  options: {
    width: number
    height: number
  }
}

export interface NetworkVisualizationFields {
  data?: NetworkVisualizationData
}

export default class NetworkVisualization extends Visualization implements NetworkVisualizationFields {
  static key: string = 'network'
  key: string = 'network'
  static title: string = 'visualizations.information.network.title'
  static description: string = 'visualizations.information.network.description'
  static longDescription: string = 'visualizations.information.network.long_description'
  static imageUrl: string = 'visualization_images/network.jpg'
  parameters: NetworkVisualizationParameters
  data?: NetworkVisualizationData

  constructor ({ id, name, progress, data }: VisualizationBaseFields & NetworkVisualizationFields) {
    super({ id, name, progress })
    this.parameters = {
      pruneLeastCitedPercentage: 5,
      bibliographicCouplingWeight: 50
    }
    this.data = data
  }
}
