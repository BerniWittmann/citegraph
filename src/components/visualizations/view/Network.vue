<template>
  <div></div>
</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator'
import * as d3 from 'd3'
import NetworkVisualization, {
  NetworkVisualizationData, NetworkVisualizationDataEdge,
  NetworkVisualizationDataNode
} from '@/models/visualizations/NetworkVisualization'

interface Coordinate {
  x: number
  y: number
}
interface Edge {
  source: Coordinate
  target: Coordinate
}

@Component
export default class Network extends Vue {
  @Prop() readonly visualization!: NetworkVisualization

  layout: any

  get diagramData (): NetworkVisualizationData {
    return this.visualization.data!
  }

  get nodes (): Array<NetworkVisualizationDataNode> {
    return this.diagramData.data.nodes
  }

  get nodesIdMap (): Record<string, NetworkVisualizationDataNode> {
    const result: Record<string, NetworkVisualizationDataNode> = {}
    this.nodes.forEach((node: NetworkVisualizationDataNode) => {
      result[node.id] = node
    })
    return result
  }

  get edges (): Array<Edge> {
    return this.diagramData.data.edges.map((edge: NetworkVisualizationDataEdge) => {
      return {
        source: this.nodesIdMap[edge.source],
        target: this.nodesIdMap[edge.target]
      }
    })
  }

  mounted () {
    const svg = d3.select(this.$el).append('svg')
      .attr('width', this.diagramData.options.width)
      .attr('height', this.diagramData.options.height)

    const g = svg.append('g')

    g.selectAll('circle .node')
      .data(this.nodes)
      .enter()
      .append('svg:circle')
      .attr('class', 'node')
      .attr('cx', (d: NetworkVisualizationDataNode) => d.x)
      .attr('cy', (d: NetworkVisualizationDataNode) => d.y)
      .attr('r', (d: NetworkVisualizationDataNode) => d.weight + 'px')
      .style('fill', (d: NetworkVisualizationDataNode) => d.color)
      .style('stroke', (d: NetworkVisualizationDataNode) => d.color)
      .style('stroke-width', 2)
      .style('fill-opacity', '0.8')

    g.selectAll('.line')
      .data(this.edges)
      .enter()
      .append('line')
      .attr('class', 'line')
      .attr('x1', (d: Edge) => d.source.x)
      .attr('y1', (d: Edge) => d.source.y)
      .attr('x2', (d: Edge) => d.target.x)
      .attr('y2', (d: Edge) => d.target.y)
      .style('stroke', 'black')
      .style('stroke-opacity', 0.1)
  }
}
</script>
