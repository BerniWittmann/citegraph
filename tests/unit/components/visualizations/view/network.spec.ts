import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../../setupPlugins'

import NetworkVisualization from '@/models/visualizations/NetworkVisualization'
import Network from '@/components/visualizations/view/Network.vue'

jest.useFakeTimers()

jest.mock('d3', () => {
  return {
    select: jest.fn().mockReturnThis(),
    append: jest.fn().mockReturnThis(),
    attr: jest.fn(function (attr, handler) {
      if (typeof handler === 'function') {
        handler({
          weight: 12,
          x: 0,
          y: 19,
          color: '#FF0000',
          source: {
            x: 0,
            y: 19
          },
          target: {
            x: 24,
            y: 54
          }
        })
      }
      // @ts-ignore
      return this
    }),
    selectAll: jest.fn().mockReturnThis(),
    data: jest.fn().mockReturnThis(),
    enter: jest.fn().mockReturnThis(),
    style: jest.fn(function (attr, handler) {
      if (typeof handler === 'function') {
        handler({
          weight: 12,
          x: 0,
          y: 19,
          color: '#FF0000',
          source: {
            x: 0,
            y: 19
          },
          target: {
            x: 24,
            y: 54
          }
        })
      }
      // @ts-ignore
      return this
    })
  }
})

describe('components/visualizations/view/Network.vue', () => {
  const getWrapper = (vis: NetworkVisualization) => {
    return shallowMount(Network, {
      i18n,
      mocks: {
        $t: (key: string) => key
      },
      propsData: {
        visualization: vis
      }
    })
  }

  const visualization = new NetworkVisualization({
    id: '1',
    name: 'My Network',
    progress: 1,
    data: {
      data: {
        nodes: [{
          id: '1',
          color: '#FF0000',
          x: 12,
          y: 24,
          weight: 24
        }, {
          id: '2',
          color: '#FF0000',
          x: 12,
          y: 24,
          weight: 24
        }, {
          id: '3',
          color: '#FF0000',
          x: 12,
          y: 24,
          weight: 24
        }],
        edges: [{
          source: '1',
          target: '2'
        }, {
          source: '3',
          target: '2'
        }]
      },
      options: {
        width: 700,
        height: 300
      }
    }
  })

  it('renders', () => {
    const wrapper = getWrapper(visualization)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
