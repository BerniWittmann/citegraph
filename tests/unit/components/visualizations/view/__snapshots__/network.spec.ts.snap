// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/visualizations/view/Network.vue renders 1`] = `<div></div>`;
